src/main/java/org/jenkinsci/plugins/scm_filter/GitHubPullRequestSkipTrait.java

```diff
@@ -0,0 +1,113 @@
+import jenkins.scm.api.trait.SCMBuilder;
+import jenkins.scm.api.trait.SCMSourceContext;
+import jenkins.scm.api.trait.SCMSourceRequest;
+
+import org.jenkinsci.plugins.github_branch_source.GitHubSCMBuilder;
+import org.jenkinsci.plugins.github_branch_source.GitHubSCMSourceRequest;
+import org.jenkinsci.plugins.github_branch_source.PullRequestSCMHead;
+
+import org.kohsuke.github.GHPullRequest;
+import org.kohsuke.stapler.DataBoundConstructor;
+
+import javax.annotation.Nonnull;
+import java.io.IOException;
+import java.util.Iterator;
+
+
+public class GitHubPullRequestSkipTrait extends SCMSourceTrait {
+
+
+    @DataBoundConstructor
+    public GitHubPullRequestSkipTrait () {
+    }
+
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    protected void decorateContext (SCMSourceContext<?, ?> context) {
+        context.withFilter(new GitHubPullRequestSkipTrait.ExcludePRsSCMHeadFilter());
+    }
+
+
+    /**
+     * Our descriptor.
+     */
+    @Extension
+    @SuppressWarnings("unused") // instantiated by Jenkins
+    public static class PullRequestSkipTraitDescriptorImpl extends SCMSourceTraitDescriptor {
+
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        @Nonnull
+        public String getDisplayName () {
+            return "Pull request title filtering behaviour";
+        }
+
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public boolean isApplicableToSCM (@Nonnull SCMDescriptor<?> scm) {
+            return scm instanceof GitSCM.DescriptorImpl;
+        }
+
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public boolean isApplicableToBuilder (@Nonnull Class<? extends SCMBuilder> builderClass) {
+            return GitHubSCMBuilder.class.isAssignableFrom(builderClass);
+        }
+    }
+
+
+    /**
+     * Filter that excludes pull requests according to its PR title (if it contains [ci skip] or [skip ci], case insensitive).
+     */
+    public static class ExcludePRsSCMHeadFilter extends SCMHeadFilter {
+
+
+        ExcludePRsSCMHeadFilter () {
+        }
+
+
+        @Override
+        public boolean isExcluded (
+                @Nonnull SCMSourceRequest scmSourceRequest,
+                @Nonnull SCMHead scmHead
+        ) throws IOException, InterruptedException {
+
+            if (scmHead instanceof PullRequestSCMHead) {
+                Iterable<GHPullRequest> pulls = ((GitHubSCMSourceRequest) scmSourceRequest).getPullRequests();
+                Iterator<GHPullRequest> pullIterator = pulls.iterator();
+                String prName = scmHead.getName();
+
+                while (pullIterator.hasNext()) {
+                    GHPullRequest pull = pullIterator.next();
+
+                    if (prName.equals("PR-" + pull.getNumber())) {
+                        String title = pull.getTitle().toLowerCase();
+                        return title.contains("[ci skip]") || title.contains("[skip ci]");
```

Comment:
How does this work?
You iterate the PRs and when you found one that indicates a skip you immediately stop the whole iteration. What happens to other PRs?

Reply:
It's the other way around. The function iterates through all PRs in GitHub and looks for the one that corresponds to the one the function is checking. Once it is found, then the loop stops and `title.contains("[ci skip]") || title.contains("[skip ci]")` is returned.

Reply:
Wait, if we already have the PR number, we still need to go through all PRs? Isn't that really inefficient? We can't just process the PR we need?

Reply:
^ 
> if we already have the PR number, we still need to go through all PRs

my concerns...

Reply:
I looked around, and I don't see any other way to do it. Also, it's very unlikely that we'll have a thousand PRs open at once on a single repository, so I don't think time complexity will be an issue. Besides, the checking done per iteration is just a single `.equal` call.

If you're curious, here are the classes involved:
* [PullRequestSCMHead](https://github.com/jenkinsci/github-branch-source-plugin/blob/master/src/main/java/org/jenkinsci/plugins/github_branch_source/PullRequestSCMHead.java), the pull request being checked
* [GitHubSCMSourceRequest](https://github.com/jenkinsci/github-branch-source-plugin/blob/master/src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSourceRequest.java), the one where I got the list of pull requests.

Reply:
We have close to 3k PR (closed), does this only look at open PR? If so I guess its not too bad. but I think we can get PR using this?

https://github.com/jenkinsci/github-branch-source-plugin/blob/e8c6c778f3aa0b67808910c9c87080052336ce73/src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSource.java#L1145
```
ghRepository = ((GitHubSCMSourceRequest) scmSourceRequest).getRepository();
ghRepository.getPullRequest(number)
```

Reply:
Actually, is this PR associated with the branch? I guess debugging in Java would tell us more detail on what data we are dealing with here.

(Then we can add more detail in the comment on why we would handle the issue using the code)

Reply:
1. Yes, after investigating, I can confirm that it only looks at open PRs.
2. OK. I will try that.
3. Nope, PRs are not associated with branches.

Reply:
@pandafox Yup that works. Done applying to the code.